src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs

```diff
@@ -107,644 +93,557 @@ public void Emit()
+                    }
+                    else
+                    {
+                        GenerateFastPathFuncForDictionary(writer, serializeMethodName, typeGenerationSpec);
+                    }
+                }
+
+                return CompleteSourceFileAndReturnText(writer);
             }
 
-            private string GenerateFastPathFuncForEnumerable(TypeGenerationSpec typeGenerationSpec)
+            private void GenerateFastPathFuncForEnumerable(SourceWriter writer, string serializeMethodName, TypeGenerationSpec typeGenerationSpec)
             {
                 Debug.Assert(typeGenerationSpec.CollectionValueType != null);
-
                 TypeGenerationSpec valueTypeGenerationSpec = _typeIndex[typeGenerationSpec.CollectionValueType];
-                string? writerMethodToCall = GetWriterMethod(valueTypeGenerationSpec);
 
-                string iterationLogic;
-                string valueToWrite;
+                GenerateFastPathFuncHeader(writer, typeGenerationSpec, serializeMethodName);
 
+                writer.WriteLine($"{WriterVarName}.WriteStartArray();");
+                writer.WriteLine();
+
+                string getCurrentElementExpr;
                 switch (typeGenerationSpec.CollectionType)
                 {
                     case CollectionType.Array:
-                        iterationLogic = $"for (int i = 0; i < {ValueVarName}.Length; i++)";
-                        valueToWrite = $"{ValueVarName}[i]";
+                        writer.WriteLine($"for (int i = 0; i < {ValueVarName}.Length; i++)");
+                        getCurrentElementExpr = $"{ValueVarName}[i]";
                         break;
+
                     case CollectionType.IListOfT:
                     case CollectionType.List:
                     case CollectionType.IList:
-                        iterationLogic = $"for (int i = 0; i < {ValueVarName}.Count; i++)";
-                        valueToWrite = $"{ValueVarName}[i]";
+                        writer.WriteLine($"for (int i = 0; i < {ValueVarName}.Count; i++)");
+                        getCurrentElementExpr = $"{ValueVarName}[i]";
                         break;
+
                     default:
                         const string elementVarName = "element";
-                        iterationLogic = $"foreach ({valueTypeGenerationSpec.TypeRef.FullyQualifiedName} {elementVarName} in {ValueVarName})";
-                        valueToWrite = elementVarName;
+                        writer.WriteLine($"foreach ({valueTypeGenerationSpec.TypeRef.FullyQualifiedName} {elementVarName} in {ValueVarName})");
+                        getCurrentElementExpr = elementVarName;
                         break;
                 };
 
-                if (valueTypeGenerationSpec.PrimitiveTypeKind is JsonPrimitiveTypeKind.Char)
-                {
-                    valueToWrite = $"{valueToWrite}.ToString()";
-                }
+                writer.WriteLine('{');
```

Comment:
Why not let the writer handle brackets and indentation?

Reply:
There are a few reasons, one I'd prefer if the writer class remained language agnostic, secondly you might want to indent without inserting brackets (e.g. loop containing a single statement), thirdly in some cases it makes sense to make non-trivial changes to indentation, e.g. here:

https://github.com/dotnet/runtime/pull/86526/files#diff-b7da6b90947be2bdebc35f367d9dc2c5bb66dd3be881af8a2d6cbb9c18108a7fR1010-R1012